### ISSUE TITLE
Lesson 5: Back propagation; Accelerated SGD; Neural net from scratch

### ISSUE THREAD

**Issue by datalass1** [datalass1/fastai#29]
### Overview
In lesson 5 we put all the pieces of training together to understand exactly what is going on when we talk about back propagation. We'll use this knowledge to create and train a simple neural network from scratch.

We'll also see how we can look inside the weights of an embedding layer, to find out what our model has learned about our categorical variables. This will let us get some insights into which movies we should probably avoid at all costs…

Although embeddings are most widely known in the context of word embeddings for NLP, they are at least as important for categorical variables in general, such as for tabular data or collaborative filtering. They can even be used with non-neural models with great success.

Great notes: https://forums.fast.ai/t/deep-learning-lesson-5-notes/31298

**Comment by datalass1** [comment]
# Lesson 5: Back propagation; Accelerated SGD; Neural net from scratch
Foundations of Neural Nets!
You need to use deep learning for computer vision for good results.
Understanding more about regularisation to sort out underfitting.

There are only 2 types of layer: parameters and activations. 
![](https://github.com/datalass1/fastai/blob/master/images/ccn_art.PNG?raw=true)
The parameters are what you model learns.
The yellow is our weight tensors/matrix, these numbers are calculated. 
Activations come from matrix multiplications (MM) and the activation funcions (in blue): an element wise function - a 20 long vector to 20 long activation. ReLU is what we normally use. 
MM followed by ReLU stacked together results is these amazing mathematical property called the **universal approximation theorem** which is if you have big enough weight matrices and enough of them it can solve any arbitrarily complex mathematical function to any arbitrarily high level of accuracy, assuming you can train the parameters time and compute. 

**Back propagation** parameters -= lr*paramets.grad

Resnet34 has 1000 columns because there is 1000 images to classify in ImageNet. This weight matrix is no good in transfer learning. You have new categories to predict. So when you train a new cnn you delete the last layer and add new weight matrices (as big as you need it to be) with a new ReLU inbetween.

**Comment by datalass1** [comment]
#### Entity Embeddings of Categorical Variables (1hr07mins)
[Cheng Gus and Felix Berkhahn](https://arxiv.org/pdf/1604.06737.pdf)
Discovering geography:
![](https://github.com/datalass1/fastai/blob/master/images/entity_embeddings.png?raw=true)

#### Weight Decay (1hr13mins)
With too many parameters the model is overfitting and not generalising well. 
Why can't I have lot's of parameters? You can. It's a fiction that too many parameters is bad.
More parameters mean more non-linearity, like life. 
One way to penalise over complexity: sum up the square of the parameters. 
In the loss function add the sum of the square of the parameters and multiple this by some number that WE choose...the weight decay.
So what should that number be? 0.1. However the default is 0.01 which is conservative. Every learner has a wd in kwargs. 

#### Stochastic Gradient Descent
Look at lesson-sgd nb.
